### ISSUE TITLE
spirv-tools does not build with gcc 4.8 (default for steam-runtime)

### ISSUE THREAD

**Issue by danginsburg** [KhronosGroup/SPIRV-Tools#855]
I get several instances of the following internal compiler error attempting to build spirv-tools with gcc 4.8:

```
In file included from /data/data/source2_main/src/thirdparty/glslang/External/spirv-tools/source/util/parse_number.h:23:0,
                 from /data/data/source2_main/src/thirdparty/glslang/External/spirv-tools/source/util/parse_number.cpp:15:
/data/data/source2_main/src/thirdparty/glslang/External/spirv-tools/source/util/hex_float.h: In instantiation of 'typename other_T::uint_type spvutils::HexFloat<T, Traits>::getRoundedNormalizedSignificand(spvutils::round_direction, bool*) [with other_T = spvutils::HexFloat<spvutils::FloatProxy<spvutils::Float16>, spvutils::HexFloatTraits<spvutils::FloatProxy<spvutils::Float16> > >; T = spvutils::FloatProxy<float>; Traits = spvutils::HexFloatTraits<spvutils::FloatProxy<float> >; typename other_T::uint_type = short unsigned int]':
/data/data/source2_main/src/thirdparty/glslang/External/spirv-tools/source/util/hex_float.h:580:40:   required from 'void spvutils::HexFloat<T, Traits>::castTo(other_T&, spvutils::round_direction) [with other_T = spvutils::HexFloat<spvutils::FloatProxy<spvutils::Float16>, spvutils::HexFloatTraits<spvutils::FloatProxy<spvutils::Float16> > >; T = spvutils::FloatProxy<float>; Traits = spvutils::HexFloatTraits<spvutils::FloatProxy<float> >]'
/data/data/source2_main/src/thirdparty/glslang/External/spirv-tools/source/util/hex_float.h:815:51:   required from here
/data/data/source2_main/src/thirdparty/glslang/External/spirv-tools/source/util/hex_float.h:490:63: internal compiler error: in dependent_type_p, at cp/pt.c:19496
             : negatable_left_shift<num_throwaway_bits>::val(1u);
                                                               ^
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-4.8/README.Bugs> for instructions.
```

gcc-4.8 is currently the default compiler for steam-runtime (https://github.com/ValveSoftware/steam-runtime)

**Comment by dneto0** [comment]
Sending to @AWoloszyn who wrote the hex_float support.   I believe he already worked around several other compiler issues.

**Comment by dneto0** [comment]
Hi @danginsburg can you be more specific on which gcc you're using?  The steam-runtime page isn't quite specific enough to differentiate for me.

My default is:
$ gcc --version
gcc (Ubuntu 4.8.4-2ubuntu1~14.04.3) 4.8.4
$ gcc-4.8 --version
gcc-4.8 (Ubuntu 4.8.4-2ubuntu1~14.04.3) 4.8.4

**Comment by danginsburg** [comment]
(steamrt_scout_amd64)~$ gcc --version
gcc (Ubuntu 4.8.1-2ubuntu1~12.04+steamrt2+srt1) 4.8.1

**Comment by greg-lunarg** [comment]
FWIW, I now have a spirv-tools branch with a new option SPIRV_OPT_ONLY which disables building of all parts of spirv-tools not needed by glslang. This essentially disables building of all pieces of spirv-tools which needs hex_float.h.

While the main goal was to disable instantiation of hex_float.h to workaround this problem, this does have the additional benefit of decreasing the size and build time of glslang when spirv-tools is included.

This option does not completely obviate this bug, but it may reduce its urgency. I am waiting to hear from DanG to see if this allows him to move forward and will post here when I hear.

I am open to a better name for this option. SPIRV_GLSLANG_BUILD?

**Comment by greg-lunarg** [comment]
I will let @AWoloszyn weigh in on the difficulty of fixing hex_float.h with gcc 4.8. I will keep the SPIRV_OPT_ONLY branch as an option if it turns out it is the best solution for DanG.

**Comment by AWoloszyn** [comment]
I will take another stab at trying to work around bugs in GCC. Unfortunately as I change this, sometimes it causes other issues to arise in other compilers.

Glslang hacks around this, but has to disable some compiler warnings in order for it to build "cleanly" which is what this was trying to avoid in the first place. I will see if there is some other formulation that satisfies the different compilers.

**Comment by dneto0** [comment]
> SPIRV_OPT_ONLY
That's a reasonable option name.

I've been musing about how to refactor SPIRV-Tools code into layers that let you just get the bits you want.  This is one instance.   But I haven't written down a plan for that (and don't have the staff either, soo...)

**Comment by dneto0** [comment]
Fixed by #858 
Please reopen if that does not work for you.
